### PR TITLE
HW02 by Kate Fedotova

### PR DESCRIPTION
**Pull request by fedotovakate02**

### CODE REVIEW

File: HW01/Kate Fedotova/kate_fedotova_hw_01_t_01.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+def angle
+  h = 3.0
+  m = 15.0
+  (h + (m / 60)) * 30 - m * 6
+end
+puts angle
```
Comment (vladhilko) [inline HW01/Kate Fedotova/kate_fedotova_hw_01_t_01.rb]: Please, remove the old files with HW01

File: HW02/Kate_Fedotova/kate_fedotova_hw_02_t_01.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+log_file = File.open('task_1.txt'){|file| file.read}
```
Comment (vladhilko) [inline HW02/Kate_Fedotova/kate_fedotova_hw_02_t_01.rb]: Why do you need this line? 🤔 You don't use it

File: HW02/Kate_Fedotova/kate_fedotova_hw_02_t_01.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    abort 'File is not exist!'
+end
+ 
+if File.open('task_1.txt')
+puts 'File open:' + 'task_1.txt'
+end
+
+file = File.new('task_1.txt')
+
+lines = file.readlines
```
Comment (vladhilko) [inline HW02/Kate_Fedotova/kate_fedotova_hw_02_t_01.rb]: you can use one line syntax 
```suggestion
lines = File.readlines('task_1.txt')
```

File: HW02/Kate_Fedotova/kate_fedotova_hw_02_t_02.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  abort 'File is not exist!'
+end
+
+logs = File.readlines('task_2.txt')
+  
+def new_format(lines)
+  array = []
+  lines.each do |line|
+    if line.match(/^\d{1,3}\.\d{1,3}\.\d*\.\d* \- \- \[\d.\/+\w*\/\d*\:\d.\:\d.\:\d. \+\d*\] \"\w* \/\w*\/\d\/\w* \w*\/\d\.\d\"/)
+      ip = line.match(/\d{1,3}\.\d{1,3}\.\d*\.\d*/)
```
Comment (vladhilko) [inline HW02/Kate_Fedotova/kate_fedotova_hw_02_t_02.rb]: We can move the same regular expressions to constant 
```suggestion
  IP = '\d{1,3}\.\d{1,3}\.\d*\.\d*'
  lines.each do |line|
    if line.match(/^#{IP} \- \- \[\d.\/+\w*\/\d*\:\d.\:\d.\:\d. \+\d*\] \"\w* \/\w*\/\d\/\w* \w*\/\d\.\d\"/)
      ip = line.match(/#{IP}/)
```

File: HW02/Kate_Fedotova/kate_fedotova_hw_02_t_01.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+puts 'File open: task_1.txt' if File.open('task_1.txt')
```
Comment (vladhilko) [inline HW02/Kate_Fedotova/kate_fedotova_hw_02_t_01.rb]: You don't need this line

File: HW02/Kate_Fedotova/kate_fedotova_hw_02_t_01.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+
+abort 'File is not exist!' unless File.exist?('task_1.txt')
+
+lines = File.readlines('task_1.txt')
+
+lines.each do |line|
+  puts line if line.include?('error')
+end
```
Comment (vladhilko) [inline HW02/Kate_Fedotova/kate_fedotova_hw_02_t_01.rb]: You need to return only the first line with an error

File: HW02/Kate_Fedotova/kate_fedotova_hw_02_t_02.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+end
+
+def get_formatted_log(line)
+  ip = line.match(/#{IP}/)
+  date = line.match(/#{DATE}/)
+  messege = line.match(/#{MESSEGE}/)[0].upcase
+  "#{date} FROM: #{ip} TO:#{messege}"
+end
+
+puts new_format(logs)
```
Comment (vladhilko) [inline HW02/Kate_Fedotova/kate_fedotova_hw_02_t_02.rb]: very readable 👍

File: HW02/Kate_Fedotova/kate_fedotova_hw_02_t_03.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+def get_events_interval(lines)
+  array = []
+  valid_events = lines.select { |line| line.include?('Calling core with') }
+  return [0] if valid_events.size <= 1
+
+  valid_events.each_cons(2) do |first, second|
+    first_log = first.match(/^\d*\-\d.\-\d. \d.\:\d.\:\d.\.\d/)[0]
+    second_log = second.match(/^\d*\-\d.\-\d. \d.\:\d.\:\d.\.\d/)[0]
+    array.push Time.parse(second_log) - Time.parse(first_log)
```
Comment (vladhilko) [inline HW02/Kate_Fedotova/kate_fedotova_hw_02_t_03.rb]: 👍

File: HW02/Kate_Fedotova/kate_fedotova_hw_02_t_03.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+abort 'File is not exist!' unless File.exist?('task_2.txt')
+
+logs = File.readlines('task_3.txt')
+
+def get_events_interval(lines)
+  array = []
+  valid_events = lines.select { |line| line.include?('Calling core with') }
+  return [0] if valid_events.size <= 1
+
+  valid_events.each_cons(2) do |first, second|
```
Comment (vladhilko) [inline HW02/Kate_Fedotova/kate_fedotova_hw_02_t_03.rb]: Could you try using `.map` here?
```suggestion
  valid_events.each_cons(2).map do |first, second|
```

Comment (fedotovakate02) [reply HW02/Kate_Fedotova/kate_fedotova_hw_02_t_03.rb]: done)

File: HW02/Kate_Fedotova/kate_fedotova_hw_02_t_03.rb [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+def get_events_interval(lines)
+  array = []
+  valid_events = lines.select { |line| line.include?('Calling core with') }
+  return [0] if valid_events.size <= 1
+
+  valid_events.each_cons(2).map do |first, second|
+    first_log = first.match(/^\d*\-\d.\-\d. \d.\:\d.\:\d.\.\d/)[0]
+    second_log = second.match(/^\d*\-\d.\-\d. \d.\:\d.\:\d.\.\d/)[0]
+    array.push Time.parse(second_log) - Time.parse(first_log)
```
Comment (vladhilko) [inline HW02/Kate_Fedotova/kate_fedotova_hw_02_t_03.rb]: now you can remove `array` at all 
```suggestion
    Time.parse(second_log) - Time.parse(first_log)
```

Comment (fedotovakate02) [reply HW02/Kate_Fedotova/kate_fedotova_hw_02_t_03.rb]: cool)

**Pull request merged**